src/common/rabbitmq/rabbitmq.service.ts

```diff
@@ -1,112 +1,135 @@
-import amqp, { Connection, Channel } from "amqplib/callback_api";
+import {
+  connect as amqplibConnect,
+  Options,
+  ChannelModel,
+  Channel,
+} from "amqplib";
 import {
   Injectable,
   Logger,
   OnModuleDestroy,
   OnApplicationShutdown,
+  OnModuleInit,
 } from "@nestjs/common";
 import { ConfigService } from "@nestjs/config";
 
+function OnError(action: "throw" | "log" = "throw"): MethodDecorator {
+  return function (
+    target: object,
+    propertyKey: string | symbol,
+    descriptor: PropertyDescriptor,
+  ) {
+    const originalMethod = descriptor.value;
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    descriptor.value = async function (...args: any[]) {
+      try {
+        return await originalMethod.apply(this, args);
+      } catch (error) {
+        if (action === "log") {
+          Logger.error(
+            (this as RabbitMQService)._error +
+              JSON.stringify((error as Error).message),
+            "RabbitMQService",
+          );
+        } else {
+          throw new Error((this as RabbitMQService)._error, { cause: error });
+        }
+      }
+    };
+    return descriptor;
+  };
+}
+
 /**
  * Service for publishing messages to a RabbitMQ queue.
  */
 @Injectable()
-export class RabbitMQService implements OnModuleDestroy, OnApplicationShutdown {
-  private connectionOptions: amqp.Options.Connect;
-  private connection: Connection;
+export class RabbitMQService
+  implements OnModuleInit, OnModuleDestroy, OnApplicationShutdown
+{
+  private connection: ChannelModel;
   private channel: Channel;
+  private static readonly configKeys = [
+    "hostname",
+    "port",
+    "username",
+    "password",
+  ] as const;
+  private connectionOptions: Required<
+    Pick<
+      Options.Connect,
+      (typeof RabbitMQService.configKeys)[number] | "protocol"
+    >
+  >;
+  _error: string;
 
-  constructor(private readonly configService: ConfigService) {
-    Logger.log("Initializing RabbitMQService.", "RabbitMQService");
-
-    const hostname = this.configService.get<string>("rabbitMq.hostname");
-    const port = this.configService.get<number>("rabbitMq.port");
-    const username = this.configService.get<string>("rabbitMq.username");
-    const password = this.configService.get<string>("rabbitMq.password");
+  constructor(private readonly configService: ConfigService) {}
 
-    if (!hostname || !port || !username || !password) {
-      Logger.error(
-        "RabbitMQ is enabled but missing one or more config variables.",
-        "RabbitMQService",
```

Comment:
I prefer the old code, as you can see the required fields immediately, but I'm not against having it programmatically either